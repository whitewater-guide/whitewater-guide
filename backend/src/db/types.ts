export interface RawTimestamped {
  created_at: Date;
  updated_at: Date;
}

/**
 * This is result of graphql query
 * It will convert Dates to strings
 */
export interface Timestamped {
  createdAt: string;
  updatedAt: string;
}

export const timestampResolvers = {
  createdAt: (resource: RawTimestamped) => resource.created_at.toISOString(),
  updatedAt: (resource: RawTimestamped) => resource.updated_at.toISOString(),
};
