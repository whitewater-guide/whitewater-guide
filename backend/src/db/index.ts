import db from './db';

export default db;
export * from './updatedAtTrigger';
export * from './types';
export * from './db';
