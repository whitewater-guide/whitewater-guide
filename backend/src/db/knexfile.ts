import * as knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: 'db',
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/development',
  },
  debug: process.env.DATABASE_DEBUG === 'true',
} as any;

const test: knex.Config = {
  client: 'pg',
  connection: {
    host: '127.0.0.1', // <------ Tests are run on host machine, not inside docker!
    port: 5432,
    database: 'test',
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/test',
  },
  debug: false, // process.env.DATABASE_DEBUG === 'true',
} as any;

const configs: { [env: string]: knex.Config } = {
  development,
  test,
};

export default configs;
