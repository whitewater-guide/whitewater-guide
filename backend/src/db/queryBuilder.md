# Query builder

Query builder is used to build deeply nested sql queries to match graphql queries

# How it works

![Schema](./docs/queryBuilder.jpg)

Consider following example graphql query:
```$graphql
query {
    sources {
        id
        name
        regions { // <--- connection to regions
            nodes {
                id
                name
            }
            count
        }
    }
}
```

1. Start building SQL query with some table. This is `root query`
2. Walk query tree and determine root field names. Some of them are simple columns, 
they are added to root query's `SELECT` clause. See `getPrimitives`
3. Others are connection fields. For each connection field `attachConnection` function runs.
As a result:
    1. CTE to select connection nodes is added to root query. This CTE is a `root query itself`.
    `build` property of connection map is used to build this CTE. *(2 on the picture)*
    2. Connection field is added to `SELECT` clause of `root query` as a subquery
    alongside primitive columns.
    This subquery is build with `buildConnectionColumn` function. *(3 on the picture)*
        1. In turn, this connection column subquery is composed of two parts
        2. The first part is another, inner CTE, to select and count nodes from outer CTE.
        Using `join` property of connection map *(4 on the picture)*, inner CTE is filtered to get connected to 
        table from `root query`. This is done by simply adding `WHERE` clause for one-to-many connections
        or by joining junction table and adding `WHERE` clause for many-to-many connections.
        It can also paginate (via `LIMIT/OFFSET`).
        3. The connection column is produced by building json from inner CTE.
        This is done in `buildConnectionJSONQuery` function. *(5 on the picture)*
4. (Not on the picture) Some others are one-to-one relations, sometimes indirect, like `section`->`region`. 
For them we want to build queries like
    ```$sql
    WITH rivers_ref as (
      SELECT * from rivers_view
      WHERE rivers_view.language = 'en'
    )
    SELECT
      sections_view.*,
      (SELECT to_json(rivers_ref.*) from rivers_ref where rivers_ref.id = sections_view.river_id) as river
    from sections_view
    WHERE sections_view.language = 'en'
    ```
    
    This is done in `attachOneToOnes` function 
5. As a result, a massive nested sql query is produced *(1 on the picture)*