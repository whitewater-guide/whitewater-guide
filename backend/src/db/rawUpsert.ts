import Knex = require('knex');

export async function rawUpsert<T>(db: Knex, query: string, queryName?: string): Promise<T> {
  const key: string = queryName || (query.split('SELECT ')[1]).split('(')[0];
  const result = await db.raw(query);
  return result.rows[0][key][0];
}
