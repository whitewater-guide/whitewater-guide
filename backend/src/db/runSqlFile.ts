import * as fs from 'fs';
import Knex = require('knex');

export async function runSqlFile(db: Knex, path: string) {
  const upsertRegion = await fs.readFileSync(path).toString();
  await db.schema.raw(upsertRegion);
}
