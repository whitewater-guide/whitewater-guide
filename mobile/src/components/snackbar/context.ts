import noop from 'lodash/noop';
import React from 'react';

import { SnackbarMessage } from './types';

export const SnackbarContext = React.createContext<SnackbarMessage | null>(
  null,
);
export const SetSnackbarContext = React.createContext<
  (value: SnackbarMessage | Error | null) => void
>(noop);
