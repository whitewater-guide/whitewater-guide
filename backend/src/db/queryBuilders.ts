import { GraphQLResolveInfo } from 'graphql';
import * as Knex from 'knex';
import { snakeCase } from 'lodash';
import { Context } from '../apollo';
import graphqlFields = require('graphql-fields');
import db from './db';

type ColumnMap<T> = {
  [field in keyof T]?: (context?: Context) => any;
};

type ConnectionsMap<T> = {
  [field in keyof T]?: { build: Builder, join: Joiner } | null;
};

type Builder = (options: Partial<QueryBuilderOptions<any>>) => Knex.QueryBuilder;
type Joiner = (table: string, query: Knex.QueryBuilder) => Knex.QueryBuilder;

export const getPrimitives = <T>(
  topLevelFields: Array<keyof T>,
  prefix: string,
  context?: Context,
  connections?: Array<keyof T>,
  customMap?: ColumnMap<T>,
): string[] => {
  return topLevelFields.reduce((result, field) => {
    // connection types and __typename are ignored
    if (field === '__typename' || connections && connections.includes(field)) {
      return result;
    }
    // Custom map allows to conditionally drop some statements based on context
    if (customMap && field in customMap) {
      const mapped = customMap[field]!(context);
      return mapped ?  [...result, `${prefix}.${mapped}`] : result;
    }
    return [...result, `${prefix}.${snakeCase(field)}`];
  }, []);
};

export interface QueryBuilderOptions<T> {
  info?: GraphQLResolveInfo;
  fieldsTree?: {[key: string]: any};
  context: Context,
  table: string;
  tableAlias?: string;
  customFieldMap?: ColumnMap<T>;
  connections?: ConnectionsMap<T>;
  knex?: Knex;
  language?: string;
  orderBy?: string;
  id?: string;
}

export const buildRootQuery = <T>(options: QueryBuilderOptions<T>): Knex.QueryBuilder => {
  const {
    info,
    fieldsTree,
    context,
    table,
    tableAlias,
    customFieldMap = {},
    connections = {},
    knex = db(),
    language = 'en',
    orderBy = 'name',
    id,
  } = options;
  const alias = tableAlias || table;
  let result = knex.from(tableAlias ? `${table} AS ${tableAlias}` : table);
  if (!fieldsTree && !info) {
    throw new Error('Provide either resolver info or parsed fields tree from resolver info');
  }
  const rootFieldsTree = fieldsTree || graphqlFields(info!);
  const topLevelFields: Array<keyof T> = Object.keys(rootFieldsTree) as any;
  const primitiveColumns = getPrimitives<T>(
    topLevelFields,
    alias,
    context,
    Object.keys(connections) as any,
    customFieldMap,
  );
  result = result.select(primitiveColumns);
  Object.entries(connections).forEach(([connectionName, value]) => {
    if (value && rootFieldsTree[connectionName]) {
      const { build, join } = value;
      attachConnection({
        knex,
        query: result,
        build,
        join,
        context,
        language,
        name: connectionName,
        fieldsTree: rootFieldsTree[connectionName],
      });
    }
  });
  if (id) {
    result = result.where(`${alias}.id`, id);
  }
  if (language) {
    result = result.where(`${alias}.language`, language);
  }
  if (orderBy) {
    result = result.orderBy(`${alias}.${orderBy}`);
  }
  return result;
};

/**
 * This will build a knex query which selects json object with fields 'nodes' (optional) and 'count' required
 * 'nodes' will contain json array of all rows (this will also include redundant count in each item)
 * @param {string} table - CTE which has 'count' column made with window function
 * @param {boolean} includeNodes - include nodes or just count them?
 * @param {Knex.Transaction | Knex} knex
 * @returns {Knex.QueryBuilder}
 */
export const buildConnectionSubquery = (table: string, includeNodes = true, knex = db()) => {
  // Assumption: we can request count without nodes, but if we request nodes we also must request count
  const nodes = includeNodes ? knex.select(knex.raw(`json_agg(${table}.*)`)).from(table) : null;
  const count = includeNodes ?
    knex.select(`${table}.count`).from(table).limit(1) : // count comes from window function column
    knex.count(`${table}.*`).from(table);
  const parts = [`'count'`, `(${count.toQuery()})`];
  if (nodes) {
    parts.push(`'nodes'`, `(${nodes.toQuery()})`);
  }
  return knex.select(knex.raw(`json_build_object(${parts.join(', ')})`));
};

export interface ConnectionBuilderOptions {
  table: string;
  join: Joiner;
  includeNodes?: boolean;
  knex?: Knex;
  limit?: number;
  offset?: number;
}

export const buildConnectionField = (options: ConnectionBuilderOptions) => {
  const {
    table,
    join,
    includeNodes = true,
    knex = db(),
    limit,
    offset,
  } = options;
  const result = buildConnectionSubquery(`${table}_internal`, includeNodes, knex);
  return result.with(`${table}_internal`, (db2) => {
    let cte = db2.select(`${table}.*`, knex.raw(`count(${table}.*) OVER()`)).from(table);
    cte = join(table, cte);
    if (limit) {
      cte = cte.limit(limit);
      if (offset) {
        cte = cte.offset(offset);
      }
    }
    return cte;
  });
};

interface AttachConnectionOptions {
  query: Knex.QueryBuilder;
  context: Context;
  name: string;
  build: Builder;
  join: Joiner;
  language: string;
  fieldsTree: {[key: string]: any};
  knex?: Knex;
}

export const attachConnection = (options: AttachConnectionOptions) => {
  const {
    query,
    language,
    context,
    name,
    build,
    join,
    fieldsTree,
    knex,
  } = options;
  const connection = buildConnectionField({
    knex,
    table: `${name}_connection`,
    join,
    includeNodes: !!fieldsTree.nodes,
  });
  return query.with(`${name}_connection`, knex!.raw(build({
      fieldsTree: fieldsTree.nodes || { id: {} },
      context,
      language,
      knex,
    }).toQuery()),
  ).select(knex!.raw(`(${connection.toQuery()}) as ${name}`));
};
