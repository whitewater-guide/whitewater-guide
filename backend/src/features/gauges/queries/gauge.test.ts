import { holdTransaction, rollbackTransaction } from '../../../db';
import { userContext } from '../../../test/context';
import { noTimestamps, runQuery } from '../../../test/db-helpers';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const query = `
  query gaugeDetails($id: ID, $language: String){
    gauge(id: $id, language: $language) {
      id
      name
      code
      levelUnit
      flowUnit
      requestParams
      cron
      url
      enabled
      createdAt
      updatedAt
      source {
        id
        language
        name
        script
        harvestMode
      }
    }
  }
`;

test('should return gauge', async () => {
  const result = await runQuery(query, { id: 'aba8c106-aaa0-11e7-abc4-cec278b6b50a' }, userContext);
  expect(result.errors).toBeUndefined();
  expect(noTimestamps(result.data!.gauge)).toMatchSnapshot();
});

test('should return null when id not specified', async () => {
  const result = await runQuery(query, { }, userContext);
  expect(result.errors).toBeUndefined();
  expect(result.data!.gauge).toBeNull();
});

test('should be able to specify language', async () => {
  const result = await runQuery(query, { id: 'aba8c106-aaa0-11e7-abc4-cec278b6b50a', language: 'ru' }, userContext);
  expect(result.errors).toBeUndefined();
  expect(result.data!.gauge.name).toBe('Галисийская линейка 1');
});

test('should be able to get basic attributes without translation', async () => {
  const result = await runQuery(query, { id: 'aba8c106-aaa0-11e7-abc4-cec278b6b50a', language: 'pt' }, userContext);
  expect(result.errors).toBeUndefined();
  expect(result.data!.gauge.name).toBe('Gauge gal1');
  expect(result.data!.gauge.url).toBe('http://ya.ru');
});

test.skip('it should return location', () => {
});

test.skip('it should return last level/flow/timestamp', () => {
});
