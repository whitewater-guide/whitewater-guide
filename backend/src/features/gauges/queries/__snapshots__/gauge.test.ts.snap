// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return gauge 1`] = `
Object {
  "code": "gal1",
  "cron": null,
  "enabled": false,
  "flowUnit": "m3/s",
  "id": "aba8c106-aaa0-11e7-abc4-cec278b6b50a",
  "levelUnit": "cm",
  "name": "Galicia gauge 1",
  "requestParams": Object {
    "foo": "bar",
  },
  "source": Object {
    "harvestMode": "allAtOnce",
    "id": "6d0d717e-aa9d-11e7-abc4-cec278b6b50a",
    "language": "en",
    "name": "Galicia",
    "script": "galicia",
  },
  "url": "http://ya.ru",
}
`;
